Exercises

ciaranightingale opened the pull request:
Solution to exercise 1. 
run WordCounter and enter file.txt as provided below when prompted. should produce a file called text.txt which contains the solution.
[file.txt](https://github.com/mantidproject/newstarter/files/3450630/file.txt)

Solution to exercise 2
run shapes.exe and it should print the ShapeSorter functionality to the terminal

Commit:
Completed functionality for E1

Need to tidy and comment code, bare bones code is now fully functional but not efficient.

Commit:
double to int in map and comments

Added comments to code and converted all the doubles to integers.

Commit:
Remove hyphens

hyphen function updated to account for multiple hyphens

hsautessella commented on exercises/ciara_nightingale/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,118 @@
+}
+
+// determine which pair is greater
+bool compare(std::pair<std::string, int> const pair_1,
```
This function name could be more descriptive, `compare` could be doing any number of things. Something more specific would improve readability when the function is called somewhere you can't see its docstring. 

Naming can be hard, and this is a tricky example, but maybe something along the lines of `isGreater` or `pairValueIsGreater`

hsautessella commented on exercises/ciara_nightingale/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,118 @@
+//clean the words, count them and put them into the map
+void processWord(std::string& word, std::map<std::string, int>& counter, std::string::size_type& maxlen, std::string::size_type const min_word_length) {
+		clean_word(word);
+		maxlen = std::max(maxlen, word.size()); //find the longest word.
```
To minimise the number of operations we're doing, this line could be moved inside the `if` block below. We don't care about the length of words we're not adding to our counter

ciaranightingale replied:
Correct me if I am wrong but doesn't the clean word function change the length of the string and so moving it inside the if statement would mean that three letter long words with punctuation would accidentally get through.

hsautessella replied:
You're right, you definitely want the `clean_word` function outside the `if` statement. I was referring to line 50 where you're finding the max word length.

ciaranightingale replied:
That makes more sense! Thank you.

hsautessella commented on exercises/ciara_nightingale/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,118 @@
+//counts the number of unique words no. of instances
+void uniqueWordCounter(std::string::size_type const min_word_length) {
+  std::map<std::string, int> counter;
+  std::string path;
```
I would put variable declarations as close to the first usage of that variable as possible (but don't declare it inside a loop). This saves people scrolling up and down to find out what type a variable is, but more importantly, it stops you declaring variables that may not be used later.

For example, you've declared `counter` and `maxlen` here, but if the file reading fails, you'll never use them and the program has reserved the memory space for nothing.

Commit:
Created Classes to Organise Shapes

created a Shape parent class and specific child classes for the separate shapes. Created methods to calculate the perimeter and area of the shapes and a Shape Sorter class to organise a vector of shapes

Commit:
e02 completed functionality

added a sub category within rectangle to deal with squares. Created a vector of shapes within the main to test functionality and functionality confirmed. Comments needed. Print function implemented

Commit:
Comments added

commented the exercise completely

Commit:
Changes to e01 made

CHanged variable and function names so that they were consistent with the required standards

Commit:
ex01 - Updated comments

created a header file to contain declarations of functions and edited comments

Commit:
ex02 - Comments updated

Function headers updated to comply with Doxygen formatting

Commit:
Return non zero int

a non zero int is returned if the file doesn't load properly

Commit:
void to int in header file

changed the uniqueWordCounter function from void to int in the header file

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Circle.h:
```diff
@@ -0,0 +1,18 @@
+  Circle(double radius);
+	///functions to calculate and return the perimeter and areaof the circle as a value of type double
+  virtual double perimeter() const override;
+  virtual double area() const override;
```
These methods are not virtual in this class, instead they are overriding a virtual function so you do not need `virtual` before it.

alicerussell1 replied:
Well done for including `override` here

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Circle.h:
```diff
@@ -0,0 +1,18 @@
+  virtual double area() const override;
+
+private:
+  double m_radius;
```
Since the radius shouldn't change once constructed this can be made `const`. Similarly for the other shapes

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Shape.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "Shape.h"
+
+// Construct a shape if no arguments are given with type Nul and zero sides
+Shape::Shape() : m_type("Nul"), m_sides(0) {}
```
Having a shape with no type or number of sides doesn't really make sense and it isn't used so this can probably be removed

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,22 @@
+  Shape(std::string type, int sides);
+	///initialise the perimeter and area functions 
+  virtual double perimeter() const = 0;
+  virtual double area() const = 0;
```
Good use of purely virtual functions and well done for making it `const`

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,22 @@
+
+private:
+  std::string m_type;
+  int m_sides;
```
These can also be `const` as the should not be changed once the shape has been constructed

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/ShapeSorter.cpp:
```diff
@@ -0,0 +1,109 @@
+  std::cout << "TYPES"
+            << "\n";
+	// Loop through all the shapes in the vector
+  for (auto const &shape : m_shapes) {
```
Good use of a range-base for loop

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/ShapeSorter.h:
```diff
@@ -0,0 +1,29 @@
+#include "Circle.h"
+#include "Rectangle.h"
+#include "Shape.h"
+#include "Triangle.h"
```
Since the class is only using `shape` you don't need to include `Circle.h`, `Rectangle.h` or `Triangle.h`. It is best practice to only include what you use.

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Rectangle.h:
```diff
@@ -0,0 +1,21 @@
+  ///default constructor to handle situations when no arguments are given
+  Rectangle();
+	///constructor to handle squares (one side length given)
+  Rectangle(double width);
```
I can see you are treating squares as just a rectangle with one side length but it would be better to treat it as a shape of it's own.

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Square.h:
```diff
@@ -0,0 +1,5 @@
+#include "Shape.h"
+
+class Square:Shape {
+
```
If a file isn't used it doesn't need to be included (however it would be best to deal with squares as either own shape - see comment above)

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Circle.cpp:
```diff
@@ -0,0 +1,29 @@
+#include "Circle.h"
+#include "Shape.h"
```
You do not need to include `Shape.h` here as it is included in the header file already.

alicerussell1 commented on exercises/ciara_nightingale/ex02_oo_basics/src/Circle.cpp:
```diff
@@ -0,0 +1,29 @@
+
+// Constructor to initialise the circle if no arguments are given
+Circle::Circle() 
+		: Shape("Circle", 1), m_radius(0) {}
```
You have included a constructor here with no arguments. This doesn't really make sense as every circle should have a radius so we should not allow the construction it without this information. It may also cause mistakes if someone else was using this code. Since it is not used I would recommend removing it. Similarly for the other shapes

alicerussell1 reviewed (changes requested):
This is a review for exercises 2. This is a really good solution, well done! The level of documentation in this is great. I've added some specific comments below, let me know if anything doesn't make sense.

Commit:
Edits to e02

Created Square sub class and added const to member variables

alicerussell1 reviewed (approved):
Exercise 2 is now looking good 👍

martyngigg commented:
I'm going to close the PR (I'll leave the branch around) as I assume work is finished on this now.
